sys/shell: refactor tokenizer code

HendrikVE opened the pull request:
### Contribution description

With this PR I simply copied PR #12085 authored by @jcarrano. Since he left the team, I hereby offer to take care of rebasing and implement required changes.

The tokenizer (the code that breaks up the line given to the shell into strings to create argv) was quite a messy piece of code. This commit refactors it into a more traditional state-machine based parser.

This fixes the issues with quote handling exposed by the recently introduced test.

### Testing procedure

Test vectors were added to the automated test. The old code could not handle this string:

```
> echo "t\e st" "\"" '\'' a\ b
shell: incorrect quoting
> echo "\""
shell: incorrect quoting
```

The new code handles it correctly.

### Issues/PRs references
Copy of #12085

#### Dependencies:
- [x] #13195

HendrikVE commented on sys/shell/shell.c:
```diff
@@ -58,164 +58,194 @@ static void flush_if_needed(void)
+    char *writepos = readpos;
+    enum PARSE_STATE pstate = PARSE_SPACE;
+
+    for (; *readpos != '\0'; readpos++) {
```
Initially, this line was `for (bool is_wordbreak; * readpos! = '\ 0'; readpos ++, is_wordbreak = false) {`, but it resulted in a seg fault on native since `is_wordbreak` was not set to false on the next iteration. I think I remember that it worked for nrf52dk and a nucleo board (otherwise I wouldn't have pushed it onto Github, but I don't have a board at home right now to test for it). Are there differences in compiler support regarding the comma operator that I should be aware of? Or did I just misunderstood how the comma operator is working?

fjmolinas replied:
Honestly no idea, maybe @kaspar030 knows?

kaspar030 replied:
> Are there differences in compiler support regarding the comma operator that I should be aware of? Or did I just misunderstood how the comma operator is working?

the operator should work the same on all platforms, it is completely standard.

kaspar030 replied:
> Initially, this line was `for (bool is_wordbreak; * readpos! = '\ 0'; readpos ++, is_wordbreak = false) {`, but it resulted in a seg fault on native since `is_wordbreak` was not set to false on the next iteration.

hm, one issue in that line would be that `is_wordbreak` is undefined in the first iteration, right? could that have caused a segfault?

HendrikVE replied:
Thank you @kaspar030, that worked out. I just tested it on my esp32 again and it works both ways, but on native only when initializing  `is_wordbreak` explicitly to false.

HendrikVE commented:
Rebased and ready for review :)

HendrikVE commented:
Rebased to master and Murdock builds were successful.

HendrikVE commented:
Could you please also take a look at this one, @fjmolinas ? :)

fjmolinas commented:
Murdock is happy with the testing side, and locally I can reproduce the error that was before.

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -66,160 +60,195 @@
+           │  (store)│     │   ┌─[\]──┘    └──[*]────┐ │     │         │
+           └──[*]──▶┴◀[*]┘   │                     │ └[*]▶┴◀──[*]──┘
+                               ↓     ┏━━━━━━━┓       ↓
+                               ├◀[\]┨NOQUOTE┃◀─────┼◀─┐
```
```suggestion
                               ├◀[\]─┨NOQUOTE┃◀─────┼◀─┐
```

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -66,160 +60,195 @@
+                               │           │         │   │
+                               │           └─[*]─────┘   │
+                               │     ┏━━━━━━━━━━━┓       │
+                               └───▶┃NOQUOTE ESC┠──[*]──┘
```
```suggestion
                               └────▶┃NOQUOTE ESC┠──[*]──┘
```

fjmolinas commented on sys/shell/shell.c:
````diff
@@ -66,160 +60,195 @@
+ *
+ State diagram for the tokenizer:
+```
+           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
+           ↓          │   ↓            │   │           ↓  │          ↓
+  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
+  ┃DQUOTE ESC┃      ┃DQUOTE ┠───["]─>┃SPACE  ┃<─[']──┨SQUOTE ┃     ┃SQUOTE ESC┃
+  ┗━━━━━━━━┯━┛      ┗━━━━━━┯┛        ┗┯━━━━┯━┛       ┗━┯━━━━━┛     ┗━━━┯━━━━━━┛
+           │         ↑     │          │    │           │     ↑(store)  │
+           │  (store)│     │   ┌─[\]──┘    └──[*]────┐ │     │         │
+           └──[*]──▶┴◀[*]┘   │                     │ └[*]▶┴◀──[*]──┘
+                               ↓     ┏━━━━━━━┓       ↓
+                               ├◀[\]┨NOQUOTE┃◀─────┼◀─┐
+                               │     ┗━━━━━┯━┛(store)↑   │
+                               │           │         │   │
+                               │           └─[*]─────┘   │
+                               │     ┏━━━━━━━━━━━┓       │
+                               └───▶┃NOQUOTE ESC┠──[*]──┘
+                                     ┗━━━━━━━━━━━┛
````
```suggestion
           ┌───[\]────┐   ┌─────["]────┐   ┌───[']─────┐  ┌───[\]────┐
           ↓          │   ↓            │   │           ↓  │          ↓
  ┏━━━━━━━━━━┓      ┏━┷━━━━━┓        ┏━┷━━━┷━┓       ┏━━━━┷━━┓     ┏━━━━━━━━━━┓
  ┃DQUOTE ESC┃      ┃DQUOTE ┠───["]─>┃SPACE  ┃<─[']──┨SQUOTE ┃     ┃SQUOTE ESC┃
  ┗━━━━━━━━┯━┛      ┗━━━━━━┯┛        ┗┯━━━━┯━┛       ┗━┯━━━━━┛     ┗━━━┯━━━━━━┛
           │         ↑     │          │    │           │     ↑(store)  │
           │  (store)│     │   ┌─[\]──┘    └──[*]────┐ │     │         │
           └──[*]───▶┴◀─[*]┘   │                     │ └[*]▶┴◀──[*]────┘
                               ↓     ┏━━━━━━━┓       ↓
                               ├◀[\]─┨NOQUOTE┃◀──────┼─◀─┐
                               │     ┗━━━━━┯━┛(store)↑   │
                               │           │         │   │
                               │           └─[*]─────┘   │
                               │     ┏━━━━━━━━━━━┓       │
                               └────▶┃NOQUOTE ESC┠──[*]──┘
                                     ┗━━━━━━━━━━━┛
```

HendrikVE replied:
I checked this and I realized that I can't format this graph in a way such that it looks good on my terminal and my IDE at the same time. Right now it looks good for my IDE, but on the terminal it's broken. What would you suggest in this case? Add an image instead or optimize it for one of these programs?

fjmolinas replied:
Whats your IDE? By in the terminal do you mean in `vim` or something like that? For me it looks the same in `vscode`, `vim` , or `cat` the file

HendrikVE replied:
I use CLion, which is IntelliJ based. Even though a monospace font is setup by default, the black triangle character seems to be larger than the other characters, therefore the issue in these lines. I will push a fixup commit which includes a comment about this in the code.

fjmolinas commented:
The PR also seems to reduce code size on `samr21-xpro`:

- master

```
12860	    132	   2624	  15616	   3d00	/home/francisco/workspace/RIOT/tests/shell/bin/samr21-xpro/tests_shell.elf
```

-pr

```
   text	   data	    bss	    dec	    hex	filename
  12824	    132	   2624	  15580	   3cdc	/home/francisco/workspace/RIOT/tests/shell/bin/samr21-xpro/tests_shell.elf
```

HendrikVE commented:
That's a nice bonus :+1:

fjmolinas commented on tests/shell/tests/01-run.py:
```diff
@@ -54,6 +54,11 @@
+    ('echo a\\', 'shell: incorrect quoting'),
+    ('echo "', 'shell: incorrect quoting'),
+    ("echo '", 'shell: incorrect quoting'),
+    ('echo "\'" \'"\'', '"echo""\'""""'),
```
Can we add a test where there are multiple spaces between argv.

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -119,107 +140,128 @@ static void print_help(const shell_command_t *command_list)
+    int argc = 0;
+    char *readpos = line;
+    char *writepos = readpos;
+    enum PARSE_STATE pstate = PARSE_SPACE;
```
Can you add a comment that says that this is the previous iteration state.

fjmolinas commented on tests/shell/tests/01-run.py:
```diff
@@ -54,6 +54,11 @@
     ('\b\b\b\becho', '\"echo\"'),
     ('help', EXPECTED_HELP),
     ('echo a string', '\"echo\"\"a\"\"string\"'),
+    ("""echo "t\e st" "\\"" '\\'' a\ b""", '"echo""te st"""""\'""a b"'),
```
I think there missing test cases that are not clear to me how they should be handled:

- `echo asdfas"asdfasdf`
- `echo asdfas'asdfasdf`
- `echo asd"fas'asdfa"sdf`

fjmolinas replied:
Note that these where not handled before either.

fjmolinas replied:
As a note in my terminal it results in:
- `echo asdfas"asdfasdf`: handled as unterminated message
- `echo asdfas'asdfasdf`: handled as unterminated message
- `echo asd"fas'asdfa"sdf`: echoes `asdfas'asdfasdf`
- `echo asd"fasasdfa"sdf`: echoes `asdfasasdfasdf`

HendrikVE replied:
I think everything works as expected. My outputs are:
(1) echo abcdef"ghijklmn => "echo""abcdef"ghijklmn"
(2) echo abcdef'ghijklmn => "echo""abcdef'ghijklmn"
(3) echo abc"def'ghijk"lmn => "echo""abc"def'ghijk"lmn"
(4) echo abc"defghijk"lmn => "echo""abc"defghijk"lmn"

The problem for (1), (3) and (4) is that the implemented `echo` command does not care about escaping `"`. A more clear output should probably look like this:
(1) echo abcdef"ghijklmn => "echo""abcdef\"ghijklmn"
(3) echo abc"def'ghijk"lmn => "echo""abc\"def'ghijk\"lmn"
(4) echo abc"defghijk"lmn => "echo""abc\"defghijk\"lmn"
Therefore, the `echo` command needs to escape `"`. I believe I should add this.

fjmolinas replied:
> The problem for (1), (3) and (4) is that the implemented echo command does not care about escaping ". A more clear output should probably look like this:

But don't you think that (1) and (2) should be considered invalid messages? its not different from:

`echo "asdfasdf` or `"echo asdfasdf'"`

HendrikVE replied:
The difference is that the quote is in the beginning vs the quote is within the current token. In former case we are in state `PARSE_SPACE`, in the latter in state `PARSE_UNQUOTED`. Currently we can reach the quoting state only from `PARSE_SPACE`. I'm not sure what is common here, but I know that at least bash would allow us to start quoting within a token.

HendrikVE replied:
I will change the code such that there is a transition from `PARSE_UNQUOTED` to `PARSE_SINGLEQUOTE` and `PARSE_DOUBLEQUOTE` and back.

HendrikVE replied:
This way your observations will be the same for the terminal and RIOT's shell.

HendrikVE commented:
I pushed a fixup commit. Could you please have a look if the diagram looks fine now?

fjmolinas commented:
> I pushed a fixup commit. Could you please have a look if the diagram looks fine now?

It looks ok!, lets not focus on that anyway, if its a little off in some cases thats OK IMO.

HendrikVE commented:
`echo` now escapes `"` with `\"`. Test cases still need to be adjusted in case we want `echo` to work this way.

HendrikVE commented:
Now the shell should handle quoting like bash does. I had to introduce a second pstate variable to remember from where I entered into `PARSE_SINGLEQUOTE` or `PARSE_DOUBLEQUOTE`: Either from `PARSE_SPACE` or from `PARSE_UNQUOTED`. The state will be restored accordingly when leaving the quote state.

HendrikVE commented:
The state diagram is deprecated by now. I'm not sure if it's worth updating it. It won't be so clean anymore anyway because of the second state I mentioned above.

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -66,160 +60,224 @@
-                }
-            }
+enum parse_state {
+    PARSE_SPACE             = 0b000,
+
+    PARSE_UNQUOTED          = 0b001,
+    PARSE_SINGLEQUOTE       = 0b010,
+    PARSE_DOUBLEQUOTE       = 0b011,
+
+    PARSE_UNQUOTED_ESC      = 0b101,
+    PARSE_SINGLEQUOTE_ESC   = 0b110,
+    PARSE_DOUBLEQUOTE_ESC   = 0b111,
```
`error: binary constants are a GCC extension [-Werror]`, change to hex?

HendrikVE replied:
Oh ok, I didn't know that. I'll change it to hex then

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -71,18 +71,21 @@
-    PARSE_UNQUOTED_ESC,
-    PARSE_SINGLEQUOTE_ESC,
-    PARSE_DOUBLEQUOTE_ESC,
+int PARSE_ESCAPE_MASK = 0b100;
```
Why not keep this as a enum or a define? A variable does not seem needed.

HendrikVE replied:
The reason I've moved this out of the enum is that the enum definition would mix real states with an escape mask for 3 of these states. I think it should be seperated from each other, so I'll rewrite it as define.

fjmolinas commented on sys/shell/shell.c:
```diff
@@ -168,7 +171,13 @@ static void handle_input_line(const shell_command_t *command_list, char *line)
     char *writepos = readpos;
-    enum PARSE_STATE pstate = PARSE_SPACE;
+
+    enum parse_state pstate = PARSE_SPACE;
```
Using enum types seems like an overhear for something that can be covered with `uint8_t`.

HendrikVE replied:
It could also be handled by more simple data types, but then we can't collect the states like it's done with the enum. I like it to have them stick together at the 'same spot'. Is there any drawback when using enum here?

fjmolinas replied:
Depending on the arch, enum are 4byte long I think. What do you mean by collecting the states?

HendrikVE replied:
By collecting I mean that values with the same purpose, the states, are bounded in a single data structure and not spread over mutliple single definitions. For me an enum improves readability because one can see that the interior values serve the same purpose. So if a single enum entry is 4 byte it wastes 3 bytes compared to an `uint8_t` which sums up to 7*3 = 21 bytes difference. Is that worth changing it? If it's considered valuable then I wont block this change.

HendrikVE replied:
Oh, or did you really just mean the variable `pstate`?

fjmolinas replied:
Yes I just meant the type of the `pstate` variable.

HendrikVE replied:
Changed it to `uint8_t` :)

fjmolinas commented:
> Now the shell should handle quoting like bash does. I had to introduce a second pstate variable to remember from where I entered into PARSE_SINGLEQUOTE or PARSE_DOUBLEQUOTE: Either from PARSE_SPACE or from PARSE_UNQUOTED. The state will be restored accordingly when leaving the quote state.

Everything is working fine now, code size has increased though:/.

> The state diagram is deprecated by now. I'm not sure if it's worth updating it. It won't be so clean anymore anyway because of the second state I mentioned above.

I'm fine with that if its compensated with more comments!

HendrikVE commented:
While replacing the state diagram with text comments I realized that `PARSE_SPACE` may should be renamed to `PARSE_BLANK` and it should also handle tab characters like spaces, what do you think, @fjmolinas ? Right now a tab with spaces surrounding it leads to the following output: `"echo""<tab character>""hello"` which does not make sense in my opinion.

HendrikVE commented:
Master does handle tabs correctly, I'll adapt the code.

HendrikVE commented:
Tabs are processed as before now. I replaced the state diagram by a text description which I hope is clear enough. I also did some other improvements in the code.

fjmolinas commented:
For the record, code size:

- pr
![image](https://user-images.githubusercontent.com/23060007/85148209-29678780-b250-11ea-96ba-781e744a2603.png)

- master
![image](https://user-images.githubusercontent.com/23060007/85148243-34221c80-b250-11ea-8bea-6c1bb7b1911a.png)

fjmolinas commented:
I'll take a closer look at the latest changes on Monday!

fjmolinas commented:
All good from my local tests https://ci.inria.fr/ci-riot-tribe/job/build-pipeline.jk/lastSuccessfulBuild/artifact/failuresummary.md/*view*/ (the failing boards had connection issues, I re-launched the test and it worked).

fjmolinas commented:
@HendrikVE please squash once my last remaining comments are addressed!

HendrikVE commented:
I changed the comments as suggested and squashed my commits. But this time, I'm not sure about how much they should be suqashed. One possibility would be to have 2 commits `sys/shell` and `test/shell`, but then the size of the commit would be quite big. What would you suggest here?

fjmolinas commented:
> I changed the comments as suggested and squashed my commits. But this time, I'm not sure about how much they should be squashed. One possibility would be to have 2 commits `sys/shell` and `test/shell`, but then the size of the commit would be quite big. What would you suggest here?

I prefer keeping the division as you have now, I'd rather split up the different changes. So LGTM!

fjmolinas commented:
Lets see if murdock is happy.

fjmolinas commented:
Ajj seems like there is a problem with a worker again...

fjmolinas commented:
Go!

HendrikVE commented:
Awesome, thank you @fjmolinas !
